src/amd_sgpio.c

```diff
@@ -0,0 +1,762 @@
+/*
+ * AMD SGPIO LED control
+ * Copyright (C) 2019, Advanced Micro Devices, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms and conditions of the GNU General Public License,
+ * version 2, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin St - Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <libgen.h>
+#include <inttypes.h>
+#include <sys/stat.h>
+#include <sys/mman.h>
+#include <sys/sysinfo.h>
+#include <sys/file.h>
+
+#if _HAVE_DMALLOC_H
+#include <dmalloc.h>
+#endif
+
+#include "config.h"
+#include "ibpi.h"
+#include "list.h"
+#include "utils.h"
+#include "amd_sgpio.h"
+
+#define HOST_CAP_EMS	(1 << 6)
+
+#define REG_TYPE_CFG	0x00
+#define REG_TYPE_TX	0x03
+#define REG_TYPE_AMD	0xC0
+
+#define MSG_TYPE_SGPIO	0x03
+
+struct sgpio_header_register {
+	uint8_t		reserved1:4;
+	uint8_t		msg_type:4;	/* 0x03 - SGPIO */
+	uint8_t		data_size;
+	uint8_t		msg_size;
+	uint8_t		reserved2:8;
+} __attribute__ ((__packed__));
+
+struct sgpio_request_register {
+	uint8_t		frame_type;
+	uint8_t		function;
+	uint8_t		reg_type;
+	uint8_t		reg_index;
+
+	uint8_t		reg_count;
+	uint8_t		reserved[3];
+} __attribute__ ((__packed__));
+
+struct sgpio_amd_register {
+	uint8_t		initiator:1;
```

Comment:
The binary fields are not designed for mapping registers. Here I found nice explanation: 
https://stackoverflow.com/questions/25345691/bit-fields-portability
So please use bitmasks.
I know that ledmon has this kind of mapping for SCSI in smp.c. It will be changed soon.